src/libs/auth.ts

```diff
@@ -97,71 +127,193 @@ const revokeTokens = async () => {
+export type AuthTokenState = AuthTokenSuccessState | AuthTokenExpiredState | AuthTokenErrorState;
 
+/**
+ * Attempts to load an auth token.
+ * When token is successfully loaded, returns a User.
+ * When token fails to load because of an expired refresh token, returns null
+ * WHen tokens fails to load because of an error, returns false
+ * @param includeBillingScope
+ * @param popUp whether signIn is attempted with a popup, or silently in the background.
+ */
+export const loadAuthToken = async (includeBillingScope = false, popUp = false): Promise<AuthTokenState> => {
+  const oldAuthTokenMetadata: TokenMetadata = authStore.get().authTokenMetadata;
+  const oldRefreshTokenMetadata: TokenMetadata = authStore.get().refreshTokenMetadata;
+
+  // for metrics, updates number of authToken load attempts for this session
   authStore.update((state) => ({
     ...state,
-    hasGcpBillingScopeThroughB2C: includeBillingScope,
-    sessionId,
-    sessionStartTime,
     authTokenMetadata: {
-      createdAt: authTokenCreatedAt,
-      expiresAt: authTokenExpiresAt,
+      ...oldAuthTokenMetadata,
+      totalTokenLoadAttemptsThisSession: authStore.get().authTokenMetadata.totalTokenLoadAttemptsThisSession + 1,
     },
   }));
-  Ajax().Metrics.captureEvent(Events.userLogin, {
-    authProvider: user.profile.idp,
-    sessionStartTime: Utils.makeCompleteDate(sessionStartTime),
-    authTokenCreatedAt: Utils.formatTimestampInSeconds(authTokenCreatedAt),
-    authTokenExpiresAt: Utils.formatTimestampInSeconds(authTokenExpiresAt),
-    jwtExpiresAt: Utils.formatTimestampInSeconds(jwtExpiresAt),
+
+  const loadedAuthTokenState: AuthTokenState = await tryLoadAuthToken(includeBillingScope, popUp);
+
+  if (loadedAuthTokenState.status === 'success') {
+    const user: User = (loadedAuthTokenState as AuthTokenSuccessState).user;
+    const userJWT: string = user.id_token!;
+    const decodedJWT: JwtPayload = jwtDecode<JwtPayload>(userJWT);
+    const authTokenCreatedAt: number = (decodedJWT as any).auth_time; // time in seconds when authorization token was created
+    const authTokenExpiresAt: number = user.expires_at!; // time in seconds when authorization token expires, as given by the oidc client
+    const authTokenId: string = uuid();
+    const jwtExpiresAt: number = (decodedJWT as any).exp; // time in seconds when the JWT expires, after which the JWT should not be read from
+    const refreshToken: string | undefined = user.refresh_token;
+    // for future might want to take refresh token and hash it to compare to a saved hashed token to see if they are the same
+    const isNewRefreshToken = !!refreshToken && refreshToken !== oldRefreshTokenMetadata.token;
+
+    // for metrics, updates authToken metadata and refresh token metadata
+    authStore.update((state) => ({
+      ...state,
+      authTokenMetadata: {
+        ...authStore.get().authTokenMetadata,
```

Comment:
Looks like there are several places in this update that access the current state of `authStore` via `authStore.get`. The previous state of the `authStore` is available in `state`.

```suggestion
        ...state.authTokenMetadata,
```